bndtools.m2e/src/bndtools/m2e/MavenBuildRepository.java

```diff
@@ -0,0 +1,246 @@
+    }
+
+    @Override
+    public File getRoot() throws Exception {
+        return null;
+    }
+
+    @Override
+    public PutResult put(InputStream stream, PutOptions options) throws Exception {
+        return null;
+    }
+
+    @Override
+    public File get(String bsn, Version version, Map<String,String> properties, DownloadListener... listeners) throws Exception {
+        if (!inited) {
+            init();
+        }
+
+        File returnVal = null;
+
+        final Entry<IMavenProjectFacade,MavenProject> entry = bsnMap.get(bsn);
+
+        if (entry != null) {
+            final MavenProject mavenProject = entry.getValue();
+
+            final File bundleFile = new File(mavenProject.getBuild().getDirectory() + "/" + mavenProject.getBuild().getFinalName() + ".jar");
+
+            if (!bundleFile.exists()) {
+                MavenUtil.execJarMojo(mavenProject, new NullProgressMonitor());
+            }
+
+            if (bundleFile.exists()) {
+                try (Jar bundle = new Jar(bundleFile)) {
+                    if (bsn.equals(bundle.getBsn()) && version.compareTo(new Version(bundle.getVersion())) == 0) {
+                        returnVal = bundleFile;
+                    }
+                }
+            }
+        }
+
+        if (returnVal != null && returnVal.exists()) {
+            if (listeners != null && listeners.length > 0) {
+                for (DownloadListener listener : listeners) {
+                    try {
+                        listener.success(returnVal);
+                    } catch (Throwable t) {}
+                }
+            }
+        }
+
+        return returnVal;
+    }
+
+    private String getBsnFromMavenProject(MavenProject mavenProject) throws Exception {
+        final File outputDir = new File(mavenProject.getBuild().getOutputDirectory());
+
+        if (outputDir.exists()) {
+            try (Jar jar = new Jar(outputDir)) {
+                String bsn = jar.getBsn();
+
+                return bsn;
+            }
+        }
+
+        return null;
+    }
+
+    private void init() {
+        inited = true;
+
+        final IProgressMonitor monitor = new NullProgressMonitor();
+
+        for (IMavenProjectFacade projectFacade : mavenProjectRegistry.getProjects()) {
+            final IProject project = projectFacade.getProject();
+
+            try {
+                final MavenProject mavenProject = getMavenProject(projectFacade, monitor);
+
+                final String bsn = getBsnFromMavenProject(mavenProject);
+
+                if (bsn != null) {
+                    Entry<IMavenProjectFacade,MavenProject> entry = new AbstractMap.SimpleImmutableEntry<>(projectFacade, mavenProject);
+                    bsnMap.put(bsn, entry);
+                }
+            } catch (Exception e) {
+                logger.logError("Unable to get bundle symbolic name for " + project.getName(), e);
+            }
+        }
+
+        mavenProjectRegistry.addMavenProjectChangedListener(this);
```

Comment:
When we remove the plugin from the Run object in RunProjectListener (below) should we not also remove this listener? Otherwise we are only ever expanding the maven project change listeners.